### DISCUSSION TITLE
airflow looks for configuration in wrong path

### DISCUSSION THREAD

**Discussion by hacxman** [apache/airflow#25022] [Q&A]
Apache Airflow version
2.2.4

OS
Ubuntu 20.04

Versions of providers:
apache-airflow-providers-ftp    1.0.0                
apache-airflow-providers-http   1.0.0                
apache-airflow-providers-imap   1.0.0                
apache-airflow-providers-sqlite 1.0.0 

Deployment
Other

Deployment details
pip3 using ansible on local LXC instance

What happened:
I've installed airflow using pip, system wide as root. I need to use it as another user, namely airflow. But, when I try to create airflow users like (manually, logged in as airflow user; normally this would be run by ansible shell task):

$ airflow users create --username admin --firstname admin --lastname .......

It complains about PermissionError: [Errno 13] Permission denied: '/root/airflow/logs/scheduler/2022-07-13'
My question is, where does it get /root/ path from? It makes no sense.
How do I tell it to use /home/airflow/airflow/ path instead?
Can you help, please?

**Comment by alexkruc** [comment]
Airflow's default configurations reside at `$AIRFLOW_HOME`, which is set to default at `~/airflow`.
Because you installed Airflow as root, the `~` value (home folder) of root is `/root/` - here is from where the `/root/airflow` is coming :) 
Here is documentation for the matter:
https://airflow.apache.org/docs/apache-airflow/stable/howto/set-config.html#setting-configuration-options

If you want to use a different path, you need to specify a different `$AIRFLOW_HOME` variable, or to pass a different configuration using `$AIRFLOW_CONFIG` variable with the path to your `airflow.cfg`. Here is the documentation for all the env vars available - 
https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#environment-variables

But this use case means that some "issues" with your installation.. If you want Airflow to work without any issues, why not install Airflow from another user in the begin with? This can cause to some unexpected results and you should re-consider your method of installation :)

**Comment by hacxman** [comment]
Thanks a lot, i've completely forgot about AIRFLOW_HOME

**Comment by hacxman** [comment]
I'm running airflow command as airflow user, so I don't understand where does it get /root/airflow path from. Grepping /usr/local/lib/python3.8/dist-packages/airflow revealed nothing.

$ whoami
airflow

$ AIRFLOW_HOME="/home/airflow/airflow" AIRFLOW_CONFIG="/home/airflow/airflow/airflow.cfg" AIRFLOW__CORE__BASE_LOGS_FOLDER="/home/airflow/airflow/logs/" airflow db init
Unable to load the config, contains a configuration error.
Traceback (most recent call last):
  File "/usr/lib/python3.8/pathlib.py", line 1288, in mkdir
    self._accessor.mkdir(self, mode)
PermissionError: [Errno 13] Permission denied: '/root/airflow/logs/scheduler/2022-07-14'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/config.py", line 563, in configure
    handler = self.configure_handler(handlers[name])
  File "/usr/lib/python3.8/logging/config.py", line 744, in configure_handler
    result = factory(**kwargs)
  File "/usr/local/lib/python3.8/dist-packages/airflow/utils/log/file_processor_handler.py", line 47, in __init__
    Path(self._get_log_directory()).mkdir(parents=True, exist_ok=True)
  File "/usr/lib/python3.8/pathlib.py", line 1297, in mkdir
    if not exist_ok or not self.is_dir():
  File "/usr/lib/python3.8/pathlib.py", line 1422, in is_dir
    return S_ISDIR(self.stat().st_mode)
  File "/usr/lib/python3.8/pathlib.py", line 1198, in stat
    return self._accessor.stat(self)
PermissionError: [Errno 13] Permission denied: '/root/airflow/logs/scheduler/2022-07-14'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/bin/airflow", line 5, in <module>
    from airflow.__main__ import main
  File "/usr/local/lib/python3.8/dist-packages/airflow/__init__.py", line 46, in <module>
    settings.initialize()
  File "/usr/local/lib/python3.8/dist-packages/airflow/settings.py", line 495, in initialize
    LOGGING_CLASS_PATH = configure_logging()
  File "/usr/local/lib/python3.8/dist-packages/airflow/logging_config.py", line 74, in configure_logging
    raise e
  File "/usr/local/lib/python3.8/dist-packages/airflow/logging_config.py", line 69, in configure_logging
    dictConfig(logging_config)
  File "/usr/lib/python3.8/logging/config.py", line 808, in dictConfig
    dictConfigClass(config).configure()
  File "/usr/lib/python3.8/logging/config.py", line 570, in configure
    raise ValueError('Unable to configure handler '
ValueError: Unable to configure handler 'processor'

**Reply by alexkruc** [reply]
`base_logs_folder` is part of `[logging]` section, and not `[core]` as you mentioned here in your env var..
Can you try it using the correct env var?
Here is the documentation:
https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html?highlight=base_log_folder#base-log-folder

TBH IDK why it's not defaulting to the `$AIRFLOW_HOME` that you said, but this is what I meant when I said that installing from user X and running from user Y can lead to a strange behaviours :)
Did you consider re-installing Airflow using the `airflow` user, if you mean to do everything with this user?

**Reply by hacxman** [reply]
Trying this results in the same error as before.
$ AIRFLOW_HOME="/home/airflow/airflow" AIRFLOW_CONFIG="/home/airflow/airflow/airflow.cfg" AIRFLOW__LOGGING__BASE_LOG_FOLDER="/home/airflow/airflow/logs/" airflow db init

Before we switched to LXC, we used Docker and for some reason (I can't recall why) we ended up with two airflow installations, one system wide (in site-packages) and other one as user. I'd really like to avoid having two copies. So that's why I'm trying to get it running this way first. I'll run airflow webserver and scheduler as systemd service. I'd prefer to have all python packages in the same place. I still consider installing as a user but only as a last resort if this attempt fails horribly :)

**Reply by hacxman** [reply]
The environment variables did indeed work from command line. The mystery was hidden inside my airflow.cfg. I did a bit of digging in sources, added a few prints here and there and in a few minutes the PEBKAC became very obvious. It still trolls hard when running from ansible :)
Thanks a lot for your patience and help!

**Reply by alexkruc** [reply]
Good to hear that it helped! :) 
You will encounter more magic like this in the future, unfortunately.. I'm just talking from experience because I also managed an Airflow installation for a few different users 😅 
BTW - you can run a process from systemd using a different user, so it's still not a blocker of moving your Airflow installation to a specific user :) 
Also - Docker also solves this for good - but I guess you had your drawbacks, since you decided to leave this approach :)
